### PR TITLE
Add 'better sandwiches' chapter

### PR DESCRIPTION
**Pull request by feihong**
Overview:

- Variant constructors can contain arguments
- OCaml's pattern matching syntax makes your code more readable and concise
  - You can match to any level of nesting
  - There are many options for combining branches in switch expressions, e.g.
    - `| A => 42 | B => 42` becomes `| A | B => 42`
    - `| A(B) => 42 | A(C) => 42` becomes `| A(B | C) => 42`
- A valid use of wildcards in switch expressions is when you want to silence
  compiler errors while you fill in the logic for new branches in switch
  expressions
- `j` quoted string literals are similar to JavaScript's template literals, but
  they are not type safe and should be avoided in production code
- `Print.sprintf` is a type-safe function for string interpolation

### CODE REVIEW

File: docs/better-sandwiches/index.md [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+We can implement this using `Js.Array.joinWith`:
+
+<<< Item.re#to-emoji-joinwith{3-13}
+
+This touches on the issue of coding style for switch expressions. When possible,
+put shorter branches before longer branches:
+
+<<< Item.re#to-emoji-short-branch-first
+
+## `j` quoted string literal
```
Comment (feihong) [inline docs/better-sandwiches/index.md]: I know the full term is something like "quoted string literal using the quoted string identifier j" but that seemed too long, so in most places, I shortened it to "j quoted string literal". But I'm now wondering if it's more clear to write "`{j||j}` quoted string literal".

File: docs/better-sandwiches/index.md [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)
+except that they can only accept variables, not arbitrary expressions. Variables
+that appear in `j` quoted string literals are implicitly converted to strings,
+so they are not type safe and thus unsuitable for production code. See the
+[exercises at the end of this chapter](#exercises) for more details.
+
+Also note that unlike switch expressions, the `fun` syntax does not accept
+multi-line expressions in branches unless you add `{}` around them.
+
+## `Printf.sprintf`
```
Comment (feihong) [inline docs/better-sandwiches/index.md:160]: I wonder if I should add a warning about how much using `Printf.sprintf` adds to the bundle size?

File: docs/better-sandwiches/index.md (hunk 1 above)
Comment (jchavarri) [reply docs/better-sandwiches/index.md]: Yeah I think what you suggested is more clear, especially for those not familiar with what "quoted strings" is, which will prob be the case.

File: docs/better-sandwiches/index.md [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+
+`j` quoted string literals are similar to [template literals in
+JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)
+except that they can only accept variables, not arbitrary expressions. Variables
+that appear in `j` quoted string literals are implicitly converted to strings,
+so they are not type safe and thus unsuitable for production code. See the
+[exercises at the end of this chapter](#exercises) for more details.
+
+Also note that unlike switch expressions, the `fun` syntax does not accept
+multi-line expressions in branches unless you add `{}` around them.
```
Comment (jchavarri) [inline docs/better-sandwiches/index.md:149]: wow, TIL.

File: docs/better-sandwiches/index.md (hunk 2 above)
Comment (jchavarri) [reply docs/better-sandwiches/index.md:160]: Yes, that'd be great I think.

File: docs/better-sandwiches/index.md [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+
+[^1]: Technically, undeclared variables produce a warning and it's possible to
+  tell OCaml to not treat them as compilation errors. A common way to do this is
+  via the [built-in `warning`
+  attribute](https://v2.ocaml.org/manual/attributes.html#ss:builtin-attributes).
+
+[^2]: You may be wondering why a unicorn sandwich costs $80. Well, that is just
+  the price of ethically-sourced unicorn meat. All the "guests" at the
+  Sanctuarium Unicornis are pampered daily with massages, horn polishings, and
+  heirloom carrots. Once a week, a 7 kg chunk of fatty meat is extracted
```
Comment (jchavarri) [inline docs/better-sandwiches/index.md:315]: 😱

**Review by jchavarri — APPROVED** [review]
🦄

File: docs/better-sandwiches/index.md (hunk 1 above)
Comment (feihong) [reply docs/better-sandwiches/index.md]: Changed "j quoted string literal" to "{j||j} quoted string literal" in https://github.com/melange-re/melange-for-react-devs/pull/13/commits/0fd698d595fa7a9eb552ae6a160a43cf4345ee0f

File: docs/better-sandwiches/index.md (hunk 2 above)
Comment (feihong) [reply docs/better-sandwiches/index.md:160]: Added an info box about how much `Printf` adds to bundle size in https://github.com/melange-re/melange-for-react-devs/pull/13/commits/c3f6d8db0d132a10a8da5ad278b5b717e1aded91

File: docs/better-sandwiches/index.md [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+<<< Item.re#to-emoji-sprintf
+
+Unlike `{j||j}` quoted string literals, `Printf.sprintf` will not do implicit
+string conversion.
+
+::: info
+
+Be aware that using functions from the `Printf` module can add a surprising
+amount to your app's production bundle size. In the app from this chapter, the
+size of `main.js` generated by Webpack went from 144.2 kB (using
```
Comment (jchavarri) [inline docs/better-sandwiches/index.md:176]: Maybe worth mentioning if these measurements are minified / uglified and gzipped or not. E.g. 

> 144.2 kB (unminified, not compressed)

**Review by jchavarri — APPROVED** [review]

File: docs/better-sandwiches/index.md (hunk 5 above)
Comment (feihong) [reply docs/better-sandwiches/index.md:176]: Indicate that the numbers are uncompressed in https://github.com/melange-re/melange-for-react-devs/pull/13/commits/036768e9de50ae4188a1bd8eb5ff4b1ceb3ed49a

**Pull request merged**